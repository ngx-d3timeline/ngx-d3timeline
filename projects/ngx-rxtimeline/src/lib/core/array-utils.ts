export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}

export function toArray(...args: any[]): any[] {
  return args;
}
