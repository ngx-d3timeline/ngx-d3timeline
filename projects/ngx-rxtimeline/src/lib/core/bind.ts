export function bind1<TBoundArg, TUnboundArgs extends any[], TOut>(
  func: (boundArg: TBoundArg, ...unboundArgs: TUnboundArgs) => TOut,
  boundVal: TBoundArg
): (...unboundArgs: TUnboundArgs) => TOut {
  return (...unboundArgs: TUnboundArgs) => func(boundVal, ...unboundArgs);
}

export function bind2<TBoundArg1, TBoundArg2, TUnboundArgs extends any[], TOut>(
  func: (
    boundArg1: TBoundArg1,
    boundArg2: TBoundArg2,
    ...unboundArgs: TUnboundArgs
  ) => TOut,
  boundVal1: TBoundArg1,
  boundVal2: TBoundArg2
): (...unboundArgs: TUnboundArgs) => TOut {
  return (...unboundArgs: TUnboundArgs) =>
    func(boundVal1, boundVal2, ...unboundArgs);
}
