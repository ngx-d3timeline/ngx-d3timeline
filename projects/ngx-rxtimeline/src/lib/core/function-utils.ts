export function partialApply<TInput, TAppliedArgs extends any[], TOut>(
  func: (input: TInput, ...appliedArgs: TAppliedArgs) => TOut
): (...appliedArgs: TAppliedArgs) => (input: TInput) => TOut {
  return (...appliedArgs: TAppliedArgs) => (input: TInput) =>
    func(input, ...appliedArgs);
}

export function pipe<T>(
  ...args: ((...args: any[]) => T)[]
): (...t: any[]) => T {
  return (...x: any[]) =>
    args.slice(1).reduce((acc, curr) => curr(acc), args[0](...x));
}

export function clampZero(x: number) {
  return Math.max(x, 0);
}

export function subtract(x: number, y: number) {
  return x - y;
}

export function mult(x: number, y: number) {
  return x * y;
}

export const negate: (x: any) => number = mult.bind(null, -1);

export function add(x: number, y: number): number {
  return x + y;
}

export const double = mult.bind(null, 2);

export function or<T>(x: T, y: T): T {
  return x || y;
}

export function identity<T>(x: T): T {
  return x;
}
