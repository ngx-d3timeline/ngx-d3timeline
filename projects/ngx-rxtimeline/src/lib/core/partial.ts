export function partial<T1, UnboundArgs extends any[], TOut>(
  func: (x1: T1, ...unboundArgs: UnboundArgs) => TOut,
  val1: T1
): (...unboundArgs: UnboundArgs) => TOut;

export function partial<T1, T2, UnboundArgs extends any[], TOut>(
  func: (x1: T1, x2: T2, ...unboundArgs: UnboundArgs) => TOut,
  val1: T1,
  val2: T2
): (...unboundArgs: UnboundArgs) => TOut;

export function partial(...args: any[]) {
  const func = args[0];
  const boundArgs = args.slice(1);

  return (...unboundArgs: any[]) => func(...boundArgs, ...unboundArgs);
}
