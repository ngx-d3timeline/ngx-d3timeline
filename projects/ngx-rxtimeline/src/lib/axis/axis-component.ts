import { Component, Input } from '@angular/core';
import { Axis } from './axis';

@Component({
  selector: '[ngx-rxtimeline-axis]',
  template: `
    <svg:g class="axis-group" [ngClass]="axis.orientation">
      <svg:g
        *ngFor="let tickMark of axis.tickMarks"
        ngx-rxtimeline-tick-mark
        [tickMark]="tickMark"
      ></svg:g>
      <ng-container *ngIf="axis.showGridLines">
        <svg:g
          *ngFor="let gridLine of axis.gridLines"
          class="grid-line"
          ngx-rxtimeline-line
          [line]="gridLine"
        ></svg:g>
      </ng-container>
      <svg:g
        class="axis-line"
        *ngIf="axis.line"
        ngx-rxtimeline-line
        [line]="axis.line"
      ></svg:g>
    </svg:g>
  `
})
export class AxisComponent {
  @Input() axis: Axis;
}
