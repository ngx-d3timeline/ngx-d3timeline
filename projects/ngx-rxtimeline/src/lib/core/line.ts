import { Orientation } from './orientation';
import { Point, translatePointInOrientation, origin } from './point';
import { partial } from './partial';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export const createOrientedLineFromOrigin = partial(createOrientedLine, origin);

export function createOrientedLine(
  point: Point,
  length: number,
  orientation: Orientation
): Line {
  const to = translatePointInOrientation(point, length, orientation);

  return createLine(point, to);
}

export function createLine(from: Point, to: Point) {
  return {
    x1: from.x,
    y1: from.y,
    x2: to.x,
    y2: to.y
  };
}
