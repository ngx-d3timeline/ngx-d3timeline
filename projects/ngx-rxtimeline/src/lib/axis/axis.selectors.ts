import { createSelector } from '../store-lib/selector/create-selector';
import {
  selectViewTopLeft,
  selectVerticalMargin,
  selectMarginLeft
} from '../view/view.selectors';
import {
  selectResourceAxisTickMarks,
  selectTimeAxisTickMarks,
  selectTimeAxisTickValues,
  selectGetTickPosition
} from '../tick-mark/tick-mark.selector';
import { Line, createOrientedLine, createLine } from '../core/line';
import {
  selectOrientedBandScale,
  selectOrientedTimeScale,
  selectResources
} from '../scales/scale-selectors';
import { OrientedScale } from '../scales/oriented-scale';
import { Scale } from '../scales/scale-types';
import { mapValues } from '../core/transform-utils';
import {
  selectAxisShowGridLines,
  selectAxisShowAxisLine
} from '../options/selectors/axis-options.selectors';
import { partialApply } from '../core/function-utils';
import { TickMark } from '../tick-mark/tick-mark';
import { Axis, AxisType } from './axis';
import { Orientation, flipOrientation } from '../core/orientation';
import { Point } from '../core/point';
import {
  selectTimeOrientation,
  selectResourceOrientation
} from '../options/selectors/options.selectors';
import { createEnumSelector } from '../store-lib/selector/selector-utils';

const selectGetAxisLine = createSelector(
  selectViewTopLeft,
  partialApply(getAxisLine)
);

function getAxisLine(
  orientedScale: OrientedScale<Scale>,
  viewTopLeft: Point
): Line {
  return createLine(viewTopLeft, getAxisEndPoint(orientedScale, viewTopLeft));
}

function getAxisEndPoint(
  orientedScale: OrientedScale<Scale>,
  viewTopLeft: Point
): Point {
  return orientedScale.orientation === Orientation.Vertical
    ? { ...viewTopLeft, y: getRangeLimit(orientedScale.scale) }
    : { ...viewTopLeft, x: getRangeLimit(orientedScale.scale) };
}

const selectGetMargin = createEnumSelector<Orientation, number>({
  Horizontal: selectMarginLeft,
  Vertical: selectVerticalMargin
});

const selectGetResourceAxisGridLine = createSelector(
  selectGetTickPosition,
  selectOrientedBandScale,
  selectOrientedTimeScale,
  selectGetMargin(selectTimeOrientation),
  partialApply(getTickGridLine)
);

const selectGetTimeAxisGridLine = createSelector(
  selectGetTickPosition,
  selectOrientedTimeScale,
  selectOrientedBandScale,
  selectGetMargin(selectResourceOrientation),
  partialApply(getTickGridLine)
);

function getTickGridLine(
  tickValue: any,
  tickPosition: (o: Orientation, range: number) => Point,
  orientedScale: OrientedScale<Scale>,
  otherOrientedScale: OrientedScale<Scale>,
  margin: number
): Line {
  return createOrientedLine(
    tickPosition(orientedScale.orientation, orientedScale.scale(tickValue)),
    getGridLineLength(otherOrientedScale, margin),
    flipOrientation(orientedScale.orientation)
  );
}

function getGridLineLength(
  orientedScale: OrientedScale<Scale>,
  margin: number
) {
  return getRangeLimit(orientedScale.scale) - margin;
}

function getRangeLimit(scale: Scale): number {
  return scale.range()[1];
}

const selectResourceAxisGridLines = createSelector(
  selectResources,
  selectGetResourceAxisGridLine,
  mapValues
);

const selectTimeAxisGridLines = createSelector(
  selectTimeAxisTickValues,
  selectGetTimeAxisGridLine,
  mapValues
);

const selectResourceAxisLine = createSelector(
  selectAxisShowAxisLine(AxisType.Resources),
  selectOrientedBandScale,
  selectGetAxisLine,
  axisLineFromOrientedScale
);

const selectTimeAxisLine = createSelector(
  selectAxisShowAxisLine(AxisType.Time),
  selectOrientedTimeScale,
  selectGetAxisLine,
  axisLineFromOrientedScale
);

export const selectResourceAxis = createSelector(
  selectResourceAxisLine,
  selectResourceAxisTickMarks,
  selectAxisShowGridLines(AxisType.Resources),
  selectResourceAxisGridLines,
  selectResourceOrientation,
  getAxis
);

export const selectTimeAxis = createSelector(
  selectTimeAxisLine,
  selectTimeAxisTickMarks,
  selectAxisShowGridLines(AxisType.Time),
  selectTimeAxisGridLines,
  selectTimeOrientation,
  getAxis
);

function getAxis(
  line: Line,
  tickMarks: TickMark[],
  showGridLines: boolean,
  gridLines: Line[],
  orientation: Orientation
): Axis {
  return {
    line,
    tickMarks,
    showGridLines,
    gridLines,
    orientation
  };
}

function axisLineFromOrientedScale(
  showAxisLine: boolean,
  orientedScale: OrientedScale<Scale>,
  line: (o: OrientedScale<Scale>) => Line
) {
  return showAxisLine && line(orientedScale);
}
