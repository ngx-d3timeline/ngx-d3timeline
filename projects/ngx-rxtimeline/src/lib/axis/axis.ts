import { TickMark } from '../tick-mark/tick-mark';
import { Line } from '../core/line';
import { Orientation } from '../core/orientation';

export interface Axis {
  tickMarks: TickMark[];
  line: Line;
  showGridLines: boolean;
  gridLines: Line[];
  orientation: Orientation;
}

export enum AxisType {
  Time = 'Time',
  Resources = 'Resources'
}
