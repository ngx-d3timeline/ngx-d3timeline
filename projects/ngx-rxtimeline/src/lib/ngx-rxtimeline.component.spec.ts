import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NgxrxtimelineComponent } from './ngx-rxtimeline.component';
import { of } from 'rxjs';
import { Component, Input } from '@angular/core';
import { Axis } from './axis/axis';
import { NgxrxtimelineService } from './ngx-rxtimeline.service';
import { View } from './view/view';
import { ResourceRectangle } from './resource-rectangle/resource-rectangle';

Object.defineProperty(window, 'ResizeObserver', {
  value: jest.fn(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn()
  }))
});

@Component({
  selector: '[ngx-rxtimeline-axis]',
  template: `
    <svg:g class="axis-group"></svg:g>
  `
})
class FakeAxisComponent {
  @Input() axis: Axis;
}

@Component({
  selector: '[ngx-rxtimeline-content]',
  template: `
    <svg:g class="content"></svg:g>
  `
})
class FakeContentComponent {}

@Component({
  selector: '[ngx-rxtimeline-resource-rectangle]',
  template: `
    <svg:rect></svg:rect>
  `
})
class FakeResourceRectangleComponent {
  @Input() resourceRectangle: ResourceRectangle;
}

describe('NgxrxtimelineComponent', () => {
  let component: NgxrxtimelineComponent;
  let fixture: ComponentFixture<NgxrxtimelineComponent>;
  let timeline: NgxrxtimelineService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        NgxrxtimelineComponent,
        FakeContentComponent,
        FakeAxisComponent,
        FakeResourceRectangleComponent
      ],
      providers: [
        {
          provide: NgxrxtimelineService,
          useValue: {
            timeAxis$: jest.fn(),
            resourceAxis$: jest.fn(),
            content$: jest.fn(),
            lastDraggedActivity$: jest.fn(),
            resourceRectangles$: jest.fn(),
            resourceTickMarkRectangles$: jest.fn(),
            setupZoom: jest.fn(),
            setupResizing: jest.fn(),
            onActivityDropped: jest.fn()
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NgxrxtimelineComponent);
    timeline = fixture.debugElement.injector.get(NgxrxtimelineService);
    timeline.activityDropped$ = of(null);

    component = fixture.componentInstance;
  });

  it('should not render if view null', () => {
    timeline.view$ = of(null);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  describe('view not null', () => {
    beforeEach(() => {
      const mockRectangles: ResourceRectangle[] = [
        {
          id: 'resource1',
          width: 10,
          height: 10,
          transform: 'translate(50,50',
          hovered: true,
          selected: true
        },
        {
          id: 'resource2',
          width: 10,
          height: 10,
          transform: 'translate(50,50',
          hovered: false,
          selected: false
        }
      ];

      timeline.view$ = of({ width: 800, height: 600 });
      timeline.resourceAxis$ = of(null);
      timeline.timeAxis$ = of(null);
      timeline.resourceRectangles$ = of(mockRectangles);
      timeline.resourceTickMarkRectangles$ = of(mockRectangles);
    });

    it('should render correctly', () => {
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });
});
