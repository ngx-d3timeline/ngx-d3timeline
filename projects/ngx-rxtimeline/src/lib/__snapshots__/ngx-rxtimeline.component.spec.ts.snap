// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxRxtimelineComponent should not render if view null 1`] = `
<div
  id="root0"
>
  
</div>
`;

exports[`NgxRxtimelineComponent view not null should render correctly 1`] = `
<div
  id="root1"
>
  
  <svg
    class="ngx-rxtimeline"
    height="600"
    width="800"
  >
    <g
      class="time-axis"
      ngx-rxtimeline-axis=""
    >
      <g
        class="axis-group"
      />
    </g>
    
    <g
      class="resource-rectangle"
      ngx-rxtimeline-resource-rectangle=""
    >
      <rect />
    </g>
    <g
      class="resource-rectangle"
      ngx-rxtimeline-resource-rectangle=""
    >
      <rect />
    </g>
    
    <g
      class="resource-title-background"
      ngx-rxtimeline-resource-rectangle=""
    >
      <rect />
    </g>
    <g
      class="resource-title-background"
      ngx-rxtimeline-resource-rectangle=""
    >
      <rect />
    </g>
    <g
      class="resources-axis"
      ngx-rxtimeline-axis=""
    >
      <g
        class="axis-group"
      />
    </g>
    <g
      ngx-rxtimeline-content=""
    >
      <g
        class="content"
      />
    </g>
  </svg>
</div>
`;
